### DISCUSSION TITLE
Task Provider - hello world example that works?

### DISCUSSION THREAD

**Discussion by chipbite** [microsoft/vscode-discussions#1577] [Extension Development QnA] [answered]
Hi!
I am frustrated since I cant get a Task Provider sample working.

This is a cross-posting, I just  posted here on so:

https://stackoverflow.com/questions/78927527/vs-code-extension-api-task-provider-build-task-example-simple

( Basically, I have tried (a) the task provider sample and there is a (b) nice hello world example posted [here on SO](https://stackoverflow.com/a/55140893/3701019), but I cant get either to work.
I put the latter in a [repo here](https://github.com/chipbite/vscode-task-provider-example). )

Very frustrated getting stumped by this seemingly simple task. Any help most appreciated!

BR!

PS
I'll put the screencapture here also for attention:

![vs-code-task-exampleProvider2](https://github.com/user-attachments/assets/8b980695-3afd-4d9b-b9f4-b5653a3596a3)

**Comment by alefragnani** [comment, marked as answer]
Hi @chipbite ,

It’s been a while I played with the Task Provider API, but I don’t remember any issues with the sample provided in the VS Code Extensions Sample repo (https://github.com/microsoft/vscode-extension-samples). Maybe you could give it a try https://github.com/microsoft/vscode-extension-samples/tree/main/task-provider-sample.

I’m not close to my computer to try out, but based on the screenshot source you provided, the only thing I’m noticing is the _lack of async characteristic_ of the `provideTasks` function, as the Task Provider API definition is:

```typescript
    registerTaskProvider(type: string, provider: TaskProvider<Task>);
```

Hope this helps

**Reply by chipbite** [reply]
Thanks alot for taking the time to read and reply! 🙇‍♂️ 🙏 

I have checked and tried the sample. It also runs fine, as the code above does - but I cannot get any of these to actually give me any tasks, always end up with `No [...] tasks found...`. 

Interesting observation on `async` in the case of `provideTasks`. 
However, I doubt that there is a problem with the signature per se, of that method, since it does indeed get called and seems to complete. Perhaps the tasks that `provideTask` returns in this example are not what/as vs code expects/wants them - but if so, vs code just fails silently - AFAICS. There is no error (again, AFAICS).

`ResolveTask` is never called, though. 

The task provider sample code of both `provideTasks` and `resolveTask` are rather different, but again since the former is indeed executing, perhaps `resolveTask` is the culprit. 

https://github.com/microsoft/vscode-extension-samples/blob/3d0af8d8c6ee4d434500adfad81386f6ea953ae0/task-provider-sample/src/rakeTaskProvider.ts#L23-L41

```ts
	public provideTasks(): Thenable<vscode.Task[]> | undefined {
		if (!this.rakePromise) {
			this.rakePromise = getRakeTasks();
		}
		return this.rakePromise;
	}


	public resolveTask(_task: vscode.Task): vscode.Task | undefined {
		const task = _task.definition.task;
		// A Rake task consists of a task and an optional file as specified in RakeTaskDefinition
		// Make sure that this looks like a Rake task by checking that there is a task.
		if (task) {
			// resolveTask requires that the same definition object be used.
			const definition: RakeTaskDefinition = <any>_task.definition;
			return new vscode.Task(definition, _task.scope ?? vscode.TaskScope.Workspace, definition.task, 'rake', new vscode.ShellExecution(`rake ${definition.task}`));
		}
		return undefined;
	}
}
```


( But I have experimented w it before, but will circle back to do it again, to make sure... )

Thanks again for your comment! BR!

**Reply by isidorn** [reply]
fyi @meganrogge

**Reply by chipbite** [reply]
I have elected to go ahead and just use the command API since it is frictionless to get along with, and using when-criteria for the filetypes I am interested in - and a bit of code - gets me pretty much what I need. I may circle back to this, since I am stubborn, but for now I am progressing without it. But again, its a scratch that I could get it to work.... :-)

**Comment by meganrogge** [comment]
based on the API docs, `resolveTask` only gets called when the task has no `execution` set, which explains why that's happening for you

https://github.com/microsoft/vscode/blob/9cca73d7e5a6d6426a91b15cedf06670929286cd/src/vscode-dts/vscode.d.ts#L8824-L8842
